src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,449 @@
+				       uint32_t align)
+{
+	/* allocate new buffer */
+	buffer->addr = (uintptr_t)rballoc_align(0, SOF_MEM_CAPS_DMA, size, align);
+
+	if (!buffer->addr) {
+		tr_err(&lib_manager_tr, "dma_buffer_init(): alloc failed");
+		return -ENOMEM;
+	}
+
+	bzero((void *)buffer->addr, size);
+	dcache_writeback_region((void *)buffer->addr, size);
+
+	/* initialise the DMA buffer */
+	buffer->size = size;
+	buffer->w_ptr = buffer->addr;
+	buffer->r_ptr = buffer->addr;
+	buffer->end_addr = buffer->addr + buffer->size;
+	buffer->avail = 0;
+	tr_dbg(&lib_manager_tr, "lib_manager_dma_buffer_init(): 0x%x, 0x%x",
+			buffer->addr, buffer->end_addr);
+	return 0;
+}
+
+void lib_manager_dma_buffer_free(struct lib_manager_dma_buf *buffer)
+{
+	rfree((void *)buffer->addr);
+	memset(buffer, 0, sizeof(struct lib_manager_dma_buf));
+}
+
+/**
+ * \brief Start, init and alloc DMA and buffer used by loader.
+ *
+ * \return 0 on success, error code otherwise.
+ */
+static int lib_manager_dma_init(struct lib_manager_dma_ext *dma_ext, uint32_t dma_id)
+{
+	uint32_t addr_align;
+	int ret = IPC4_FAILURE;
+
+	/* request DMA in the dir HMEM->LMEM */
+	dma_ext->dma.dmac = dma_get(DMA_DIR_HMEM_TO_LMEM, 0, DMA_DEV_HOST,
+			DMA_ACCESS_EXCLUSIVE);
+	if (!dma_ext->dma.dmac) {
+		tr_err(&lib_manager_tr, "dma_ext_init(): dma.dmac = NULL");
+		return -ENODEV;
+	}
+
+	/* get required address alignment for dma buffer */
+	ret = dma_get_attribute(dma_ext->dma.dmac, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+					&addr_align);
+	if (ret < 0)
+		return ret;
+
+	dma_ext->chan = dma_channel_get(dma_ext->dma.dmac, dma_id);
+	if (!dma_ext->chan)
+		return -EINVAL;
+
+	return 0;
+}
+
+/**
+ * \brief Stop, deinit and free DMA and buffer used by loader.
+ *
+ * \return 0 on success, error code otherwise.
+ */
+
+static int lib_manager_dma_deinit(struct lib_manager_dma_ext *dma_ext)
+{
+	dma_channel_put(dma_ext->chan);
+	dma_put(dma_ext->dma.dmac);
+	return 0;
+}
+
+static int lib_manager_load_data_from_host(struct lib_manager_dma_ext *dma_ext, uint32_t size)
+{
+	struct dma_sg_config config;
+	uint32_t elem_addr;
+	uint32_t elem_num, elem_size;
+	int ret = IPC4_FAILURE;
+	uint32_t free_bytes = 0;
+	uint32_t avail_bytes = 0;
+
+	elem_addr = (uint32_t)dma_ext->dmabp.addr;
+
+	elem_size = sizeof(uint64_t) * DMA_ELEM_SIZE;
+	elem_num = size / elem_size;
+
+	config.direction = DMA_DIR_HMEM_TO_LMEM;
+	config.src_width = sizeof(uint32_t);
+	config.dest_width = sizeof(uint32_t);
+	config.cyclic = 0;
+
+	ret = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
+			   config.direction, elem_num, elem_size, elem_addr, 0);
+
+	if (ret < 0)
+		return ret;
+
+	ret = dma_set_config(dma_ext->chan, &config);
+	if (ret < 0)
+		return ret;
+
+	ret = dma_start(dma_ext->chan);
+	if (ret < 0)
+		return ret;
+
+	/* Wait till whole data acquired */
+	while (avail_bytes < size) {
+		/* get data sizes from DMA */
+		ret = dma_get_data_size(dma_ext->chan, &avail_bytes, &free_bytes);
```

Comment:
Whats the plan for using the Zephyr DMA APIs here ?

Reply:
@abonislawski @juimonen pls confirm if we can use native Zephyr DMA calls for all stages here.